core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -554,6 +555,32 @@ public String authority() {
     }
   }
 
+  private final class DirectFallbackExecutor implements Executor {
+    @Override
+    public void execute(final Runnable command) {
+      final boolean[] accepted = new boolean[1];
+      try {
+        ManagedChannelImpl.this.executor.execute(new Runnable() {
+          @Override
+          public void run() {
+            accepted[0] = true;
+            command.run();
+          }
+        });
+      } catch (RejectedExecutionException e) {
```

Comment:
NPE is not a concern:
```java
Executor saved = executor;
if (saved == null) {
  command.run();
} else {
  saved.execute(command);
}
```

I agree there is still a race between checking `executor` and running with it, but it is dramatically smaller and controlled by our code instead of the executor.

In my mind the race is _calling_ `executor` after the channel has terminated. None of these solutions actually fix that. They just mitigate it.